### PR TITLE
Feature/iat 583

### PR DESCRIPTION
**Pull request by hbgwhite**
Hooked up validation endpoint for items via RabbitMQ

### CODE REVIEW

File: src/main/java/org/opentestsystem/ap/ivs/utilities/StringUtilities.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.opentestsystem.ap.ivs.utilities;
+
+import java.util.regex.Pattern;
+
+/**
+ * Created by gregwhite on 8/16/17.
+ */
+public class StringUtilities {
+
+    public static boolean isNullOrEmpty(String target) {
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/utilities/StringUtilities.java]: check out https://commons.apache.org/proper/commons-lang/javadocs/api-3.4/org/apache/commons/lang3/StringUtils.html  We have this on the classpath and it is really helpful.  If I understand this method correctly it is the StringUtils.isEmtpy equivalent.  

We also have commons collection which is helpful for collections like isEmtpy(List).

File: src/main/java/org/opentestsystem/ap/ivs/config/GitLabConfig.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import org.springframework.stereotype.Component;
+
+/**
+ * Created by gregwhite on 8/16/17.
+ */
+@Getter
+@Setter
+@Component
+@ConfigurationProperties(prefix = "gitlab")
+public class GitLabConfig {
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/config/GitLabConfig.java]: I wonder if we should rename this to IvsProperties.  I like "properties" over "config" in the name.  I see config as something else like configuring spring beans.  Where classes like this are just properties

We could rename the prefix to "ivs" and renamed the property to "gitlabSecret".

I see in the ValidationService below there are some hardcoded values we could make properties in IvsProperties like validationQueue and validationTopicExchange.  I would set the values in IvsProperties class right on the class.  See the common class ItemBankProperties and notice some have default values.  That means we don't have to put the properties in application.yml.  But if we ever need/want to override the default values it is a simple matter of adding the property to the config files.

File: src/main/java/org/opentestsystem/ap/ivs/models/GitLabActionDto.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ */
+public class GitLabActionDto implements Serializable {
+    public GitLabActionDto(GitLabAction action) {
+        this.id = action.checkout_sha;
+        this.project = action.project.name;
+        this.username = action.user_name;
+        this.userEmail = action.user_email;
+    }
+
+    public String username;
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/models/GitLabActionDto.java]: Make the properties private and add the Lombok @Getter and @Setter annotations at the class level.  Or just @Getter if you want the class immutable.  Then access the properties using the getters.  This is the Java way :)  The annotations help reduce verbosity.  Java is lacking in a true "property" concept like C# has.

File: src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    TopicExchange exchange() {
+        return new TopicExchange("cpt-exchange");
+    }
+
+    @Bean
+    Binding binding(Queue queue, TopicExchange exchange) {
+        return BindingBuilder.bind(queue).to(exchange).with(QUEUE_NAME);
+    }
+
+    @Bean
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java]: I would move all these @Bean methods into a Config class under the config package.  Maybe call it IvsConfig or MessageConfig might be better making it specific to queues, topics, exchanges, etc.  Config classes annotated with @Configuration is where @Bean definitions go in the spring world.

File: src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ * Created by gregwhite on 8/15/17.
+ */
+@Slf4j
+@Component
+public class ValidationService {
+
+    final static String QUEUE_NAME = "cpt";
+
+    @Autowired
+    RabbitTemplate rabbitTemplate;
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java]: You could make this final and then autowire it via a constructor.  We prefer constructor injection or field injection.

File: src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer;
+
+/**
+ * Created by gregwhite on 8/15/17.
+ */
+@Slf4j
+@Component
+public class ValidationService {
+
+    final static String QUEUE_NAME = "cpt";
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java]: This is a value I was thinking to put in IvsProperties.

File: src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    RabbitTemplate rabbitTemplate;
+
+    @Bean
+    Queue queue() {
+        return new Queue(QUEUE_NAME, false);
+    }
+
+    @Bean
+    TopicExchange exchange() {
+        return new TopicExchange("cpt-exchange");
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java]: Another value good for IvsProperties.  And you'll want to inject IvsProperties into this service class.

Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java]: Actually you will want to inject IvsProperties into MessageConfig given that is where the bean definitions will be.

**Comment by bdrainer** [comment]
I suggested some class names like IvsProperties and MessageConfig.  If you feel there are better names then use what you feel best.  Maybe ValidationProperties and/or BrokerConfig are better.

**Comment by bdrainer** [comment]
I hit the wrong button

File: src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return container;
+    }
+
+    @Bean
+    MessageListenerAdapter listenerAdapter(AmqpReceiver receiver) {
+        return new MessageListenerAdapter(receiver, "receiveMessage");
+    }
+
+    @Bean
+    AmqpReceiver amqpReceiver() {
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java]: Is this what will consume the message on the queue and ultimately do the work of validating?

File: src/main/java/org/opentestsystem/ap/ivs/service/AmqpReceiver.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ */
+@Slf4j
+public class AmqpReceiver {
+
+    /**
+     * Receive items off of AMQP RabbitMq and prepare for validation
+     *
+     * @param message The GitLabActionDto containing the information necessary to validate the item
+     */
+    public void receiveMessage(Object message) {
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/service/AmqpReceiver.java]: Ah sorry just answered my own question that I asked below about this.

File: src/main/java/org/opentestsystem/ap/ivs/rest/v1/ValidationApi.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    // ------------------------------------------------------------------------
+    // POST requests - Validate Item
+    // ------------------------------------------------------------------------
+
+    /**
+     * Places item on AMQP queue to be validated
+     *
+     * @param gitLabAction The GitLab push hook containing item information
+     */
+    @PostMapping("")
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/rest/v1/ValidationApi.java]: you can remove the ("")  and just have @PostMapping

File: src/main/java/org/opentestsystem/ap/ivs/rest/v1/ValidationApi.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Validation on IAT items
+ */
+@Slf4j
+@RestController
+@RequestMapping("/api/v1/validation")
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/ivs/rest/v1/ValidationApi.java]: Let's talk with Ray about the urls being used.  We are attempting to be as RESTful as we can so I want his opinion on what would be the best REST urls for this.

File: src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java (hunk 8 above)
Comment (hbgwhite) [inline src/main/java/org/opentestsystem/ap/ivs/service/ValidationService.java]: Yes it is

**Comment by bdrainer** [comment]
Nice work.  Good learning on how to incorporate rabbit messaging.  Use a template to produce a message and then register a listener as a consumer.  Very cool.

Just push any changes to this branch and it will apply to this PR.

**Pull request merged**